File: clients/apple/AudiobookAppleApp/Features/Player/ViewModels/PlayerViewModel.swift
import Foundation
import AudiobookCore
import AVFoundation
import MediaPlayer
#if canImport(UIKit)
import UIKit
private typealias PlatformImage = UIImage
#elseif canImport(AppKit)
import AppKit
private typealias PlatformImage = NSImage
#endif

@MainActor
final class PlayerViewModel: ObservableObject {
    @Published private(set) var state = PlayerViewState()

    private let repository: PlayerRepository
    private let authService: AuthService
    private let realtime: RealtimeClient
    private let deviceId: String
    private let deviceLabel = "Apple App"
    private var lastClaimTime: TimeInterval = 0
    private var lastLiveProgressEmitAt: TimeInterval = 0
    private var isRealtimeBound = false
    private var presenceTask: Task<Void, Never>?
    private var autosaveTask: Task<Void, Never>?
    private var player: AVPlayer?
    private var periodicObserver: Any?
    private var remoteCommandsConfigured = false
    private var playCommandTarget: Any?
    private var pauseCommandTarget: Any?
    private var skipForwardCommandTarget: Any?
    private var skipBackwardCommandTarget: Any?
    private var artworkTask: Task<Void, Never>?
    private var lastArtworkBookId: String?

    init(repository: PlayerRepository, authService: AuthService, realtime: RealtimeClient) {
        self.repository = repository
        self.authService = authService
        self.realtime = realtime
        let defaults = UserDefaults.standard
        if let existing = defaults.string(forKey: "player_apple_device_id") {
            self.deviceId = existing
        } else {
            let generated = UUID().uuidString
            defaults.set(generated, forKey: "player_apple_device_id")
            self.deviceId = generated
        }
    }

    func load(bookId: String, title: String) async {
        bindRealtimeIfNeeded()
        state.isLoading = true
        state.errorMessage = nil
        state.bookId = bookId
        state.title = title

        do {
            async let settingsTask = repository.fetchSettings()
            async let playbackDetailsTask = repository.fetchPlaybackDetails(bookId: bookId)
            let resume = try await repository.resumeInfo(bookId: bookId)
            let dbProgress = try? await repository.fetchProgress(bookId: bookId)
            let settings = try await settingsTask
            let playbackDetails = try await playbackDetailsTask
            let authoritativePosition = Double(dbProgress?.positionSeconds ?? Int(resume.startSeconds))
            let playerSettings = settings.player
            state.streamPath = resume.streamPath
            state.streamURLString = repository.streamURL(streamPath: resume.streamPath).absoluteString
            state.startSeconds = resume.startSeconds
            state.positionSeconds = authoritativePosition
            state.durationSeconds = Double(dbProgress?.durationAtSave ?? Int(resume.durationSeconds))
            state.author = playbackDetails.author
            state.coverPath = playbackDetails.coverPath
            state.backwardJumpSeconds = Double(playerSettings.backwardJumpSeconds ?? 10)
            state.forwardJumpSeconds = Double(playerSettings.forwardJumpSeconds ?? 30)
            state.playbackRate = playerSettings.playbackRate ?? 1.0
            state.appliedRewind = resume.appliedRewind
            state.chapters = playbackDetails.chapters
            state.currentChapterIndex = chapterIndex(for: authoritativePosition)
            configurePlayerIfNeeded()
            seekPlayer(to: authoritativePosition)
            configureRemoteCommandsIfNeeded()
            updateNowPlayingInfo()
            loadNowPlayingArtworkIfNeeded()
            broadcastPresence()
        } catch {
            state.errorMessage = "Could not load playback data."
        }

        state.isLoading = false
    }

    func updatePosition(_ value: Double) {
        state.positionSeconds = value
        state.currentChapterIndex = chapterIndex(for: value)
        seekPlayer(to: value)
        updateNowPlayingInfo()
        broadcastLiveProgress(force: false)
    }

    func selectChapter(_ index: Int) {
        guard let chapter = state.chapters.first(where: { $0.index == index }) ?? state.chapters.dropFirst(index).first else {
            return
        }

        updatePosition(Double(chapter.start) / 1000)
    }

    func handleSkipBackwardMediaAction() {
        seekBy(-state.backwardJumpSeconds)
    }

    func handleSkipForwardMediaAction() {
        seekBy(state.forwardJumpSeconds)
    }

    private func seekBy(_ deltaSeconds: Double) {
        updatePosition(max(0, state.positionSeconds + deltaSeconds))
    }

    func playPressed() {
        configurePlayerIfNeeded()
        player?.playImmediately(atRate: Float(state.playbackRate))
        state.isPlaying = true
        state.activeDeviceLabel = nil
        updateNowPlayingInfo()
        claimPlaybackOwnership()
        broadcastPresence()
        startAutosaveLoop()
    }

    func pausePressed() {
        player?.pause()
        state.isPlaying = false
        updateNowPlayingInfo()
        broadcastPresence()
        stopAutosaveLoop()
        Task {
            await saveProgressSilently()
        }
    }

    func saveProgress() async {
        guard !state.bookId.isEmpty else {
            return
        }

        state.isSaving = true
        state.errorMessage = nil

        do {
            try await repository.saveProgress(
                bookId: state.bookId,
                positionSeconds: Int(state.positionSeconds),
                durationAtSave: max(Int(state.durationSeconds), 1)
            )
            broadcastLiveProgress(force: true)
        } catch {
            state.errorMessage = "Could not save progress."
        }

        state.isSaving = false
    }

    func reset() {
        player?.pause()
        state.isPlaying = false
        broadcastPresence()
        stopAutosaveLoop()
        artworkTask?.cancel()
        clearNowPlayingInfo()
        state = PlayerViewState()
    }

    deinit {
        stopAutosaveLoop()
        presenceTask?.cancel()
        if let periodicObserver {
            player?.removeTimeObserver(periodicObserver)
        }
        artworkTask?.cancel()
        removeRemoteCommandTargets()
        clearNowPlayingInfo()
        player?.pause()
        realtime.disconnect()
    }

    private func bindRealtimeIfNeeded() {
        guard !isRealtimeBound else {
            return
        }

        isRealtimeBound = true
        realtime.connect { [weak self] event in
            Task { @MainActor in
                self?.handleRealtimeEvent(event)
            }
        }

        presenceTask?.cancel()
        presenceTask = Task { [weak self] in
            while !(Task.isCancelled) {
                try? await Task.sleep(nanoseconds: 10_000_000_000)
                await MainActor.run {
                    self?.broadcastPresence()
                }
            }
        }
    }

    private func configurePlayerIfNeeded() {
        guard let streamURLString = state.streamURLString,
              let url = URL(string: streamURLString)
        else {
            return
        }

        let currentURL = (player?.currentItem?.asset as? AVURLAsset)?.url.absoluteString
        if currentURL == streamURLString {
            return
        }

        if let periodicObserver {
            player?.removeTimeObserver(periodicObserver)
            self.periodicObserver = nil
        }

        let freshPlayer = AVPlayer(url: url)
        freshPlayer.automaticallyWaitsToMinimizeStalling = true
        let observer = freshPlayer.addPeriodicTimeObserver(
            forInterval: CMTime(seconds: 1, preferredTimescale: 600),
            queue: .main
        ) { [weak self] time in
            guard let self else {
                return
            }

            let seconds = CMTimeGetSeconds(time)
            if seconds.isFinite {
                self.state.positionSeconds = max(0, seconds)
                self.state.currentChapterIndex = self.chapterIndex(for: self.state.positionSeconds)
            }

            if let duration = freshPlayer.currentItem?.duration.seconds, duration.isFinite, duration > 0 {
                self.state.durationSeconds = duration
            }

            self.updateNowPlayingInfo()
            self.broadcastLiveProgress(force: false)
        }

        self.player = freshPlayer
        self.periodicObserver = observer
        if state.isPlaying {
            freshPlayer.playImmediately(atRate: Float(state.playbackRate))
        }
        updateNowPlayingInfo()
    }

    private func seekPlayer(to seconds: Double) {
        guard let player else {
            return
        }

        let clamped = max(0, seconds)
        player.seek(to: CMTime(seconds: clamped, preferredTimescale: 600))
    }

    private func handleRealtimeEvent(_ event: RealtimeEventEnvelope) {
        guard let payload = event.payload else {
            return
        }

        switch event.type {
        case "progress.synced":
            guard payload.string("userId") == authService.userId,
                  payload.string("bookId") == state.bookId,
                  !state.isPlaying
            else {
                return
            }

            Task {
                guard let snapshot = try? await repository.fetchProgress(bookId: state.bookId) else {
                    return
                }
                await MainActor.run {
                    state.positionSeconds = Double(snapshot.positionSeconds ?? Int(state.positionSeconds))
                    if let durationAtSave = snapshot.durationAtSave, durationAtSave > 0 {
                        state.durationSeconds = Double(durationAtSave)
                    }
                    state.currentChapterIndex = chapterIndex(for: state.positionSeconds)
                    seekPlayer(to: state.positionSeconds)
                    updateNowPlayingInfo()
                }
            }
        case "playback.session.presence":
            guard payload.string("userId") == authService.userId,
                  payload.string("deviceId") != deviceId,
                  payload.bool("paused") == false
            else {
                return
            }
            state.activeDeviceLabel = payload.string("label") ?? "another device"
        case "playback.claimed":
            guard payload.string("userId") == authService.userId,
                  payload.string("bookId") == state.bookId
            else {
                return
            }

            let claimDevice = payload.string("deviceId")
            let claimTime = ISO8601DateFormatter().date(from: payload.string("timestamp") ?? "")?.timeIntervalSince1970 ?? 0
            guard claimTime >= lastClaimTime else {
                return
            }

            lastClaimTime = claimTime
            if claimDevice != deviceId && state.isPlaying {
                player?.pause()
                state.isPlaying = false
                stopAutosaveLoop()
                updateNowPlayingInfo()
            }
        default:
            break
        }
    }

    private func claimPlaybackOwnership() {
        guard let userId = authService.userId, !state.bookId.isEmpty else {
            return
        }

        let now = Date()
        lastClaimTime = now.timeIntervalSince1970
        realtime.send(
            type: "playback.claim",
            payload: [
                "userId": userId,
                "deviceId": deviceId,
                "bookId": state.bookId,
                "timestamp": ISO8601DateFormatter().string(from: now),
            ]
        )
    }

    private func broadcastPresence() {
        guard let userId = authService.userId else {
            return
        }

        realtime.send(
            type: "playback.session.presence",
            payload: [
                "userId": userId,
                "deviceId": deviceId,
                "label": deviceLabel,
                "platform": "apple",
                "currentBookId": state.bookId.isEmpty ? NSNull() : state.bookId,
                "paused": !state.isPlaying,
            ]
        )
    }

    private func broadcastLiveProgress(force: Bool) {
        guard let userId = authService.userId,
              !state.bookId.isEmpty,
              state.durationSeconds > 0
        else {
            return
        }

        let now = Date().timeIntervalSince1970
        if !force && now - lastLiveProgressEmitAt < 2 {
            return
        }

        lastLiveProgressEmitAt = now
        realtime.send(
            type: "playback.progress",
            payload: [
                "userId": userId,
                "bookId": state.bookId,
                "positionSeconds": Int(state.positionSeconds),
                "durationAtSave": Int(state.durationSeconds),
                "completed": Int(state.positionSeconds) >= max(Int(state.durationSeconds) - 1, 0),
                "timestamp": ISO8601DateFormatter().string(from: Date()),
            ]
        )
    }

    private func startAutosaveLoop() {
        autosaveTask?.cancel()
        autosaveTask = Task { [weak self] in
            while !(Task.isCancelled) {
                try? await Task.sleep(nanoseconds: 15_000_000_000)
                await self?.saveProgressSilently()
            }
        }
    }

    private func stopAutosaveLoop() {
        autosaveTask?.cancel()
        autosaveTask = nil
    }

    private func saveProgressSilently() async {
        guard !state.bookId.isEmpty, state.durationSeconds > 0 else {
            return
        }

        try? await repository.saveProgress(
            bookId: state.bookId,
            positionSeconds: Int(state.positionSeconds),
            durationAtSave: max(Int(state.durationSeconds), 1)
        )
        broadcastLiveProgress(force: true)
    }

    private func configureRemoteCommandsIfNeeded() {
        guard !remoteCommandsConfigured else {
            return
        }

        remoteCommandsConfigured = true
        let center = MPRemoteCommandCenter.shared()
        center.playCommand.isEnabled = true
        center.pauseCommand.isEnabled = true
        center.skipForwardCommand.isEnabled = true
        center.skipBackwardCommand.isEnabled = true

        center.skipForwardCommand.preferredIntervals = [NSNumber(value: state.forwardJumpSeconds)]
        center.skipBackwardCommand.preferredIntervals = [NSNumber(value: state.backwardJumpSeconds)]

        playCommandTarget = center.playCommand.addTarget { [weak self] _ in
            guard let self else {
                return .commandFailed
            }

            self.playPressed()
            return .success
        }

        pauseCommandTarget = center.pauseCommand.addTarget { [weak self] _ in
            guard let self else {
                return .commandFailed
            }

            self.pausePressed()
            return .success
        }

        skipForwardCommandTarget = center.skipForwardCommand.addTarget { [weak self] _ in
            guard let self else {
                return .commandFailed
            }

            self.handleSkipForwardMediaAction()
            return .success
        }

        skipBackwardCommandTarget = center.skipBackwardCommand.addTarget { [weak self] _ in
            guard let self else {
                return .commandFailed
            }

            self.handleSkipBackwardMediaAction()
            return .success
        }
    }

    private func updateNowPlayingInfo() {
        guard !state.bookId.isEmpty else {
            MPNowPlayingInfoCenter.default().nowPlayingInfo = nil
            return
        }

        var info: [String: Any] = [
            MPMediaItemPropertyTitle: state.title,
            MPMediaItemPropertyArtist: state.author ?? "Unknown author",
            MPNowPlayingInfoPropertyElapsedPlaybackTime: state.positionSeconds,
            MPMediaItemPropertyPlaybackDuration: max(state.durationSeconds, 0),
            MPNowPlayingInfoPropertyPlaybackRate: state.isPlaying ? state.playbackRate : 0,
        ]

        if let fallbackArtwork = fallbackNowPlayingArtwork() {
            info[MPMediaItemPropertyArtwork] = fallbackArtwork
        }

        if let chapter = state.chapters[safe: state.currentChapterIndex] {
            info[MPNowPlayingInfoPropertyChapterNumber] = chapter.index + 1
            info[MPMediaItemPropertyAlbumTitle] = chapter.title
        }

        MPRemoteCommandCenter.shared().skipForwardCommand.preferredIntervals = [NSNumber(value: state.forwardJumpSeconds)]
        MPRemoteCommandCenter.shared().skipBackwardCommand.preferredIntervals = [NSNumber(value: state.backwardJumpSeconds)]
        MPNowPlayingInfoCenter.default().nowPlayingInfo = info
    }

    private func loadNowPlayingArtworkIfNeeded() {
        guard !state.bookId.isEmpty, state.coverPath != nil else {
            return
        }

        guard lastArtworkBookId != state.bookId else {
            return
        }

        lastArtworkBookId = state.bookId
        artworkTask?.cancel()
        artworkTask = Task { [weak self] in
            guard let self else {
                return
            }

            let coverURL = repository.streamURL(streamPath: "streaming/books/\(state.bookId)/cover")
            var request = URLRequest(url: coverURL)
            if let token = authService.accessToken {
                request.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")
            }

            guard let (data, response) = try? await URLSession.shared.data(for: request),
                  let http = response as? HTTPURLResponse,
                  (200...299).contains(http.statusCode),
                  let artwork = makeNowPlayingArtwork(from: data)
            else {
                return
            }

            await MainActor.run {
                guard state.bookId == lastArtworkBookId else {
                    return
                }

                var info = MPNowPlayingInfoCenter.default().nowPlayingInfo ?? [:]
                info[MPMediaItemPropertyArtwork] = artwork
                MPNowPlayingInfoCenter.default().nowPlayingInfo = info
            }
        }
    }

    private func makeNowPlayingArtwork(from data: Data) -> MPMediaItemArtwork? {
        #if canImport(UIKit)
        guard let image = PlatformImage(data: data) else {
            return nil
        }
        return MPMediaItemArtwork(boundsSize: image.size) { _ in image }
        #elseif canImport(AppKit)
        guard let image = PlatformImage(data: data) else {
            return nil
        }
        return MPMediaItemArtwork(boundsSize: image.size) { _ in image }
        #else
        return nil
        #endif
    }

    private func fallbackNowPlayingArtwork() -> MPMediaItemArtwork? {
        #if canImport(UIKit)
        let size = CGSize(width: 256, height: 256)
        let renderer = UIGraphicsImageRenderer(size: size)
        let image = renderer.image { context in
            UIColor(red: 0.16, green: 0.23, blue: 0.35, alpha: 1).setFill()
            context.fill(CGRect(origin: .zero, size: size))
        }
        return MPMediaItemArtwork(boundsSize: image.size) { _ in image }
        #elseif canImport(AppKit)
        let size = CGSize(width: 256, height: 256)
        let image = PlatformImage(size: size)
        image.lockFocus()
        NSColor(calibratedRed: 0.16, green: 0.23, blue: 0.35, alpha: 1).setFill()
        NSBezierPath(rect: NSRect(origin: .zero, size: size)).fill()
        image.unlockFocus()
        return MPMediaItemArtwork(boundsSize: image.size) { _ in image }
        #else
        return nil
        #endif
    }

    private func clearNowPlayingInfo() {
        lastArtworkBookId = nil
        MPNowPlayingInfoCenter.default().nowPlayingInfo = nil
    }

    private func removeRemoteCommandTargets() {
        guard remoteCommandsConfigured else {
            return
        }

        let center = MPRemoteCommandCenter.shared()
        if let playCommandTarget {
            center.playCommand.removeTarget(playCommandTarget)
        }
        if let pauseCommandTarget {
            center.pauseCommand.removeTarget(pauseCommandTarget)
        }
        if let skipForwardCommandTarget {
            center.skipForwardCommand.removeTarget(skipForwardCommandTarget)
        }
        if let skipBackwardCommandTarget {
            center.skipBackwardCommand.removeTarget(skipBackwardCommandTarget)
        }
        remoteCommandsConfigured = false
    }

    private func chapterIndex(for positionSeconds: Double) -> Int {
        guard !state.chapters.isEmpty else {
            return 0
        }

        let positionMilliseconds = Int(positionSeconds * 1000)
        if let match = state.chapters.lastIndex(where: { chapter in
            positionMilliseconds >= chapter.start && (chapter.end <= chapter.start || positionMilliseconds < chapter.end)
        }) {
            return match
        }

        if let last = state.chapters.last, positionMilliseconds >= last.start {
            return state.chapters.count - 1
        }

        return 0
    }
}

private extension Array {
    subscript(safe index: Index) -> Element? {
        indices.contains(index) ? self[index] : nil
    }
}

private extension Dictionary where Key == String, Value == RealtimeJSONValue {
    func string(_ key: String) -> String? {
        guard let value = self[key] else {
            return nil
        }
        if case .string(let text) = value {
            return text
        }
        return nil
    }

    func bool(_ key: String) -> Bool? {
        guard let value = self[key] else {
            return nil
        }
        if case .bool(let flag) = value {
            return flag
        }
        return nil
    }
}

File: clients/android/app/src/main/java/com/audiobook/app/PlayerViewModel.kt
package com.audiobook.app

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.audiobook.core.domain.PlayerChapter
import com.audiobook.core.domain.PlayerRepository
import com.audiobook.core.realtime.RealtimeClient
import kotlinx.coroutines.Job
import kotlinx.coroutines.delay
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch
import org.json.JSONObject

data class PlayerUiState(
    val bookId: String = "",
    val title: String = "",
    val author: String? = null,
    val coverPath: String? = null,
    val streamPath: String? = null,
    val positionSeconds: Double = 0.0,
    val durationSeconds: Double = 0.0,
    val backwardJumpSeconds: Int = 10,
    val forwardJumpSeconds: Int = 30,
    val playbackRate: Double = 1.0,
    val appliedRewind: Boolean = false,
    val chapters: List<PlayerChapter> = emptyList(),
    val currentChapterIndex: Int = 0,
    val isPlaying: Boolean = false,
    val activeDeviceLabel: String? = null,
    val isLoading: Boolean = false,
    val isSaving: Boolean = false,
    val errorMessage: String? = null
)

class PlayerViewModel(private val repository: PlayerRepository) : ViewModel() {
    private val _state = MutableStateFlow(PlayerUiState())
    val state: StateFlow<PlayerUiState> = _state.asStateFlow()
    private var userId: String? = null
    private var deviceId: String? = null
    private var deviceLabel: String = "Android App"
    private var realtimeClient: RealtimeClient? = null
    private var lastClaimTimeMs: Long = 0
    private var presenceJob: Job? = null
    private var autosaveJob: Job? = null

    fun bindRealtime(userId: String?, deviceId: String, deviceLabel: String, client: RealtimeClient) {
        this.userId = userId
        this.deviceId = deviceId
        this.deviceLabel = deviceLabel
        this.realtimeClient = client

        client.connect { type, payload ->
            when (type) {
                "progress.synced" -> onProgressSynced(payload)
                "playback.claimed" -> onPlaybackClaimed(payload)
                "playback.session.presence" -> onPresence(payload)
            }
        }

        presenceJob?.cancel()
        presenceJob = viewModelScope.launch {
            while (true) {
                delay(10_000)
                broadcastPresence()
            }
        }
    }

    fun loadPlayer(bookId: String, title: String) {
        viewModelScope.launch {
            _state.value = _state.value.copy(
                isLoading = true,
                errorMessage = null,
                bookId = bookId,
                title = title
            )
            try {
                val settings = repository.fetchSettings()
                val resume = repository.loadResume(bookId)
                val snapshot = repository.fetchProgress(bookId)
                val details = repository.fetchPlaybackDetails(bookId)
                val authoritativePosition = snapshot?.positionSeconds?.toDouble() ?: resume.startSeconds
                val authoritativeDuration = snapshot?.durationAtSave?.takeIf { it > 0 }?.toDouble() ?: resume.durationSeconds
                _state.value = _state.value.copy(
                    isLoading = false,
                    streamPath = resume.streamPath,
                    author = details.author,
                    coverPath = details.coverPath,
                    positionSeconds = authoritativePosition,
                    durationSeconds = authoritativeDuration,
                    backwardJumpSeconds = settings.backwardJumpSeconds,
                    forwardJumpSeconds = settings.forwardJumpSeconds,
                    playbackRate = settings.playbackRate,
                    appliedRewind = resume.appliedRewind,
                    chapters = details.chapters,
                    currentChapterIndex = chapterIndexForPosition(details.chapters, authoritativePosition)
                )
                broadcastPresence()
            } catch (e: Exception) {
                _state.value = _state.value.copy(
                    isLoading = false,
                    errorMessage = "Could not load playback data."
                )
            }
        }
    }

    fun updatePosition(value: Double) {
        val chapters = _state.value.chapters
        _state.value = _state.value.copy(
            positionSeconds = value,
            currentChapterIndex = chapterIndexForPosition(chapters, value)
        )
        broadcastProgress()
    }

    fun handleSkipBackwardAction() {
        seekBy(-_state.value.backwardJumpSeconds)
    }

    fun handleSkipForwardAction() {
        seekBy(_state.value.forwardJumpSeconds)
    }

    fun selectChapter(index: Int) {
        val chapter = _state.value.chapters.getOrNull(index) ?: return
        updatePosition(chapter.startSeconds)
    }

    private fun seekBy(deltaSeconds: Int) {
        val current = _state.value
        updatePosition((current.positionSeconds + deltaSeconds).coerceAtLeast(0.0))
    }

    fun playPressed() {
        _state.value = _state.value.copy(isPlaying = true, activeDeviceLabel = null)
        claimPlaybackOwnership()
        broadcastPresence()
        startAutosaveLoop()
    }

    fun pausePressed() {
        _state.value = _state.value.copy(isPlaying = false)
        broadcastPresence()
        stopAutosaveLoop()
        saveProgressSilently()
    }

    fun syncFromEngine(positionSeconds: Double, durationSeconds: Double) {
        val current = _state.value
        if (current.bookId.isBlank()) {
            return
        }

        _state.value = _state.value.copy(
            positionSeconds = positionSeconds.coerceAtLeast(0.0),
            durationSeconds = if (durationSeconds > 0) durationSeconds else current.durationSeconds,
            currentChapterIndex = chapterIndexForPosition(current.chapters, positionSeconds)
        )
        broadcastProgress()
    }

    fun saveProgress() {
        val current = _state.value
        if (current.bookId.isBlank() || current.durationSeconds <= 0) {
            return
        }

        viewModelScope.launch {
            _state.value = _state.value.copy(isSaving = true, errorMessage = null)
            try {
                repository.saveProgress(
                    bookId = current.bookId,
                    positionSeconds = current.positionSeconds.toInt(),
                    durationAtSave = current.durationSeconds.toInt().coerceAtLeast(1)
                )
                _state.value = _state.value.copy(isSaving = false)
                broadcastProgress()
            } catch (e: Exception) {
                _state.value = _state.value.copy(
                    isSaving = false,
                    errorMessage = "Could not save progress."
                )
            }
        }
    }

    fun reset() {
        stopAutosaveLoop()
        _state.value = PlayerUiState()
    }

    private fun onProgressSynced(payload: JSONObject) {
        val current = _state.value
        val payloadUser = payload.optString("userId")
        val payloadBook = payload.optString("bookId")
        if (payloadUser != userId || payloadBook != current.bookId || current.isPlaying) {
            return
        }

        viewModelScope.launch {
            val snapshot = repository.fetchProgress(current.bookId) ?: return@launch
            _state.value = _state.value.copy(
                positionSeconds = snapshot.positionSeconds.toDouble(),
                durationSeconds = if (snapshot.durationAtSave > 0) snapshot.durationAtSave.toDouble() else current.durationSeconds,
                currentChapterIndex = chapterIndexForPosition(current.chapters, snapshot.positionSeconds.toDouble())
            )
        }
    }

    private fun onPlaybackClaimed(payload: JSONObject) {
        val payloadUser = payload.optString("userId")
        val payloadDevice = payload.optString("deviceId")
        val payloadBook = payload.optString("bookId")
        val timestamp = payload.optString("timestamp")
        val claimTime = runCatching { java.time.Instant.parse(timestamp).toEpochMilli() }.getOrDefault(0L)
        if (payloadUser != userId || payloadBook != _state.value.bookId || claimTime < lastClaimTimeMs) {
            return
        }

        lastClaimTimeMs = claimTime
        if (payloadDevice != deviceId && _state.value.isPlaying) {
            stopAutosaveLoop()
            _state.value = _state.value.copy(
                isPlaying = false,
                activeDeviceLabel = _state.value.activeDeviceLabel ?: "another device"
            )
        }
    }

    private fun onPresence(payload: JSONObject) {
        val payloadUser = payload.optString("userId")
        val payloadDevice = payload.optString("deviceId")
        val paused = payload.optBoolean("paused", true)
        val label = payload.optString("label", "another device")
        if (payloadUser != userId || payloadDevice == deviceId) {
            return
        }

        if (!paused) {
            _state.value = _state.value.copy(activeDeviceLabel = label)
        }
    }

    private fun broadcastPresence() {
        val current = _state.value
        val uid = userId ?: return
        val did = deviceId ?: return
        realtimeClient?.send(
            type = "playback.session.presence",
            payload = JSONObject()
                .put("userId", uid)
                .put("deviceId", did)
                .put("label", deviceLabel)
                .put("platform", "android")
                .put("currentBookId", if (current.bookId.isBlank()) JSONObject.NULL else current.bookId)
                .put("paused", !current.isPlaying)
        )
    }

    private fun claimPlaybackOwnership() {
        val current = _state.value
        val uid = userId ?: return
        val did = deviceId ?: return
        if (current.bookId.isBlank()) {
            return
        }

        val now = java.time.Instant.now().toString()
        lastClaimTimeMs = java.time.Instant.parse(now).toEpochMilli()
        realtimeClient?.send(
            type = "playback.claim",
            payload = JSONObject()
                .put("userId", uid)
                .put("deviceId", did)
                .put("bookId", current.bookId)
                .put("timestamp", now)
        )
    }

    private fun broadcastProgress() {
        val current = _state.value
        val uid = userId ?: return
        if (current.bookId.isBlank() || current.durationSeconds <= 0) {
            return
        }

        realtimeClient?.send(
            type = "playback.progress",
            payload = JSONObject()
                .put("userId", uid)
                .put("bookId", current.bookId)
                .put("positionSeconds", current.positionSeconds.toInt())
                .put("durationAtSave", current.durationSeconds.toInt())
                .put("completed", current.positionSeconds.toInt() >= (current.durationSeconds.toInt() - 1).coerceAtLeast(0))
                .put("timestamp", java.time.Instant.now().toString())
        )
    }

    private fun startAutosaveLoop() {
        autosaveJob?.cancel()
        autosaveJob = viewModelScope.launch {
            while (true) {
                delay(15_000)
                saveProgressSilently()
            }
        }
    }

    private fun stopAutosaveLoop() {
        autosaveJob?.cancel()
        autosaveJob = null
    }

    private fun saveProgressSilently() {
        val current = _state.value
        if (current.bookId.isBlank() || current.durationSeconds <= 0) {
            return
        }

        viewModelScope.launch {
            runCatching {
                repository.saveProgress(
                    bookId = current.bookId,
                    positionSeconds = current.positionSeconds.toInt(),
                    durationAtSave = current.durationSeconds.toInt().coerceAtLeast(1)
                )
            }
            broadcastProgress()
        }
    }

    override fun onCleared() {
        presenceJob?.cancel()
        autosaveJob?.cancel()
        realtimeClient?.disconnect()
        super.onCleared()
    }

    private fun chapterIndexForPosition(chapters: List<PlayerChapter>, positionSeconds: Double): Int {
        if (chapters.isEmpty()) {
            return 0
        }

        val normalizedPosition = positionSeconds.coerceAtLeast(0.0)
        return chapters.indexOfLast { chapter ->
            normalizedPosition >= chapter.startSeconds &&
                (chapter.endSeconds <= chapter.startSeconds || normalizedPosition < chapter.endSeconds)
        }.takeIf { it >= 0 } ?: chapters.lastIndex.coerceAtLeast(0).takeIf {
            normalizedPosition >= chapters.last().startSeconds
        } ?: 0
    }
}
